### PR TITLE
#226 Add auditing for prepare statements

### PR DESCRIPTION
**Pull request by paulchandler**
#226

### CODE REVIEW

File: integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/ITQueryLogger.java [hunk 1]
```diff
@@ -88,6 +93,28 @@ public static void afterClass()
...
+    public void testPrepareStatement()
+    {
+        givenTable("school", "students");
+        reset(mockAuditAppender);
+
+        PreparedStatement prepared = session.prepare("INSERT INTO school.students (key, value) VALUES (?, ?)");
+        session.execute(prepared.bind(42, "Kalle"));
+        assertThat(getLogEntries()).containsOnly( "client:'127.0.0.1'|user:'anonymous'|status:'ATTEMPT'|operation:'Prepared: INSERT INTO school.students (key, value) VALUES (?, ?)'",
+                                                  "client:'127.0.0.1'|user:'anonymous'|status:'ATTEMPT'|operation:'INSERT INTO school.students (key, value) VALUES (?, ?)[42, 'Kalle']'");
+
```
Comment (tommystendahl) [inline integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/ITQueryLogger.java:106]: remove empty line

File: integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/PrepareAuditQueryLogger.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import org.junit.runner.RunWith;
+import org.slf4j.LoggerFactory;
+
+import ch.qos.logback.classic.LoggerContext;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Appender;
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.PreparedStatement;
+
```
Comment (tommystendahl) [inline integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/PrepareAuditQueryLogger.java:35]: remove empty line

File: integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/PrepareAuditQueryLogger.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        givenTable("school", "students");
+        reset(mockAuditAppender);
+
+        PreparedStatement prepared = session.prepare("INSERT INTO school.students (key, value) VALUES (?, ?)");
+        session.execute(prepared.bind(42, "Kalle"));
+        assertThat(getLogEntries()).containsOnly( "client:'127.0.0.1'|user:'anonymous'|status:'ATTEMPT'|operation:'Prepared: INSERT INTO school.students (key, value) VALUES (?, ?)'",
+                                                  "client:'127.0.0.1'|user:'anonymous'|status:'ATTEMPT'|operation:'INSERT INTO school.students (key, value) VALUES (?, ?)[42, 'Kalle']'");
+    }
+    @Test
```
Comment (tommystendahl) [inline integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/PrepareAuditQueryLogger.java:103]: Add empty line above

File: integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/PrepareAuditQueryLogger.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        givenTable("school", "students");
+        reset(mockAuditAppender);
+
+        assertThatExceptionOfType(InvalidQueryException.class).isThrownBy(() ->  session.prepare("INSERT INTO school.invalidestudents (key, value) VALUES (?, ?)"));
+
+        assertThat(getLogEntries()).containsOnly( "client:'127.0.0.1'|user:'anonymous'|status:'ATTEMPT'|operation:'Prepared: INSERT INTO school.invalidestudents (key, value) VALUES (?, ?)'",
+                                                  "client:'127.0.0.1'|user:'anonymous'|status:'FAILED'|operation:'Prepared: INSERT INTO school.invalidestudents (key, value) VALUES (?, ?)'");
+    }
+
+
```
Comment (tommystendahl) [inline integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/PrepareAuditQueryLogger.java:115]: remove one empty line

**Review by tommystendahl — CHANGES_REQUESTED** [review]

File: .idea/misc.xml
[the remark this answers is not in the document]
Comment (tommystendahl) [reply .idea/misc.xml:1]: No, the file is still deleted in the PR. Its deleted in the first commit so I think you have to do a commit that puts the file back.

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/PrepareAuditOperation.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        this.operationString = "Prepared: " + operationString;
+    }
+
+    @Override
+    public String getOperationString()
+    {
+        return operationString;
+    }
+    @Override
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/entry/suppressor/PrepareAuditOperation.java:38]: You missed this one.

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/role/RoleAuditFilter.java [hunk 6]
```diff
@@ -74,6 +74,10 @@ public void setup()
         whitelistDataAccess.setup();
     }
 
+    @Override
+    public boolean shouldLogPrepareStatements() { return true; }
+
+
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/role/RoleAuditFilter.java:80]: Still two empty lines here.

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/yamlandrole/YamlAndRoleAuditFilter.java [hunk 7]
```diff
@@ -57,4 +57,10 @@ public void setup()
         yamlFilter.setup();
         roleFilter.setup();
     }
+    @Override
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/yamlandrole/YamlAndRoleAuditFilter.java:60]: and this one

File: ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/yamlandrole/YamlAndRoleAuditFilter.java [hunk 8]
```diff
@@ -57,4 +57,10 @@ public void setup()
...
+    public boolean shouldLogPrepareStatements()
+    {
+        return yamlFilter.shouldLogPrepareStatements();
+    }
+
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply ecaudit/src/main/java/com/ericsson/bss/cassandra/ecaudit/filter/yamlandrole/YamlAndRoleAuditFilter.java:65]: and this one

File: ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java [hunk 9]
```diff
@@ -169,6 +170,8 @@ public void testPrepareAndGetPreparedWhenPreloaded()
 
         verify(mockHandler, times(1)).prepare(eq(query), eq(mockQueryState), eq(customPayload));
         verify(mockHandler, times(1)).getPrepared(eq(statementId));
+        verify(mockAdapter, times(1)).auditPrepare(eq(query), eq(mockClientState), eq(Status.ATTEMPT), longThat(isCloseToNow()));
+
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply ecaudit/src/test/java/com/ericsson/bss/cassandra/ecaudit/handler/TestAuditQueryHandler.java:174]: and here

File: integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/ITQueryLogger.java [hunk 10]
```diff
@@ -30,8 +31,11 @@
 import ch.qos.logback.classic.spi.ILoggingEvent;
 import ch.qos.logback.core.Appender;
 import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.PreparedStatement;
+
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/ITQueryLogger.java:35]: and here

File: integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/ITQueryLogger.java [hunk 11]
```diff
@@ -88,6 +93,28 @@ public static void afterClass()
         session.close();
         cluster.close();
     }
+    @Test
```
[the remark this answers is not in the document]
Comment (tommystendahl) [reply integration-test-query-logger/src/test/java/com/ericsson/bss/cassandra/ecaudit/integration/querylogger/ITQueryLogger.java:96]: and here

**Review by tommystendahl — CHANGES_REQUESTED** [review]

File: .idea/misc.xml [hunk 12]
```diff
@@ -1,4 +1,3 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tommystendahl) [inline .idea/misc.xml:1]: You missed the first line in the file.

**Review by tommystendahl — CHANGES_REQUESTED** [review]

File: .idea/misc.xml (hunk 12 above)
Comment (paulchandler) [reply .idea/misc.xml]: I have added this back in now.

**Review by tommystendahl — APPROVED** [review]

**Pull request merged**